### ISSUE TITLE
`.xopp`-files with stroke point width values of 0 cause crash when rendered

### ISSUE THREAD

**Issue by bhennion** [xournalpp/xournalpp#4346]
After 553588a4 being merged some `.xopp`-files either do not open any more (when opened from the file browser) or crash when reaching a certain page. One example:

```xml
<xournal creator="xournalpp 1.1.2+dev" fileversion="4">
<title>Xournal++ document - see </title>
<page width="595.276" height="841.89">
<background type="solid" color="white" style="graph" />
<layer>
<stroke tool="pen" ts="0" fn="" color="#002e99ff" width="1.41 0.49994156 0.54505829 0.56364716 0.57535126 0.58877655 0.60151337 0.61115204 0.61771408 0.62046799 0.62184494 0.62287766 0.62322189 0.62322189 0.62322189 0.62459885 0.62632004 0.62941818 0.63251633 0.63527023 0.63733566 0.63871262 0.63940109 0.63940109 0.63871262 0.63802414 0.6376799 0.6376799 0.63802414 0.63802414 0.63802414 0.63561447 0.6376799 0.64387619 0.64731857 0.65041672 0.65282639 0.6538591 0.65489182 0.65523605 0.65592453 0.65592453 0.65558029 0.65489182 0.65420334 0.65420334 0.65454758 0.6655632 0.67416916 0.6862175 0.69757736 0.70551636 0.71102417 0.71274536 0.7130896 0.7130896 0.7130896 0.71205688 0.70930298 0.6979216 0.58017059 0" capStyle="round">348.96909 148.79162 348.69433 148.63138 348.48832 148.40244 348.39673 148.10484 348.4654 147.76144 348.69433 147.46384 349.08355 147.16624 349.63298 146.84574 350.274 146.66259 350.96082 146.47945 351.73917 146.3421 352.54045 146.25052 353.27301 146.20474 353.95983 146.18184 354.55505 146.22763 355.05874 146.3421 355.5395 146.54813 355.92868 146.84574 356.18048 147.21202 356.31786 147.5783 356.31786 147.99038 356.2034 148.37956 355.99735 148.81452 355.72263 149.27237 355.40212 149.75312 355.05874 150.30255 354.66951 150.82907 354.30326 151.30983 353.93695 151.76769 353.59352 152.06529 353.34172 152.29421 353.04409 152.50025 353.34172 152.54604 354.02849 152.52315 354.44059 152.54604 354.82977 152.68339 355.24187 152.86654 355.56238 153.14125 355.81422 153.46175 356.04314 153.75936 356.22627 154.17142 356.36365 154.62928 356.52391 155.13292 356.63837 155.63656 356.72996 156.07152 356.86729 156.43781 357.07334 156.91856 356.77575 157.05591 356.47811 156.98724 356.11181 156.89567 355.65396 156.82699 355.12741 156.68963 354.50926 156.52938 353.82249 156.39202 353.11276 156.32334 352.35727 156.27756 351.62471 156.27756 350.9379 156.32334 350.36554 156.39202 349.97637 156.4836 350.25108 156.09442 350.57159 155.8426</stroke>
</layer>
</page>
</xournal>
```
which crashes. It doesn't crash when replacing the `0` pressure value in `0.58017059 0" capStyle="round">` in the `stroke tool` line by some positive value, e.g. `0.1`

_Originally posted by @rolandlo in https://github.com/xournalpp/xournalpp/issues/4301#issuecomment-1286940527_

**Other posts from there**
@rolandlo:
> The situation is quite bad. Files I generated with my stylus before January 2021 (i.e. before minimal pressures were introduced in #2622) can't be opened anymore (they all contain some 0.0000000 pressure values, sometimes even negative pressure values). Files generated after January 2021 can be opened, with the exception of the file which made me aware of the bug. I believe that I copied some strokes from an old file (annotating basically the same PDF).
> Maybe we should take care in the `LoadHandler` of that problem (replace pressures <=0 by the minimal pressure from the settings).
>
>There are two problems with that idea though:
>
> 1. the minimum pressure can also be set to 0.00
> 2. Neither the LoadHandler nor the Stroke class have access to the settings.

**Comment by bhennion** [comment]
Here are two comments:

1. We do not want to modify the opened file based on a local settings: either a file is corrupted or is not, the user's preferences have nothing to do with this. So whatever we do must be independent of Settings.
2. There is a difference between a minimal pressure value, and a minimal point width (namely a factor by stroke->getWidth()). I guess we want to set the latter.

Then, I think there are two independent solutions for two questions:
1. Fixing opened corrupted file: we could say that if a point's pressure value is 0, we reset it to  0.5*(previousPoint.z + nextPoint.z) (if either previousPoint of nextPoint does not exist, just take half the value of the other one. If neither exist, delete the stroke).
2. Filtering input events: an input event with pressure value 0 should just be ignored in PenInputHandler.

**Comment by Technius** [comment]
Instead of asserting that the pressure is nonzero, maybe a simpler short-term solution is to skip drawing the segment if the first point is zero? Assuming no other invariants are broken, it should have the same effect.
https://github.com/xournalpp/xournalpp/blob/024c3419f1194adefa72e37e7fcdf408065d5f2b/src/core/view/StrokeViewHelper.cpp#L45-L46

```c++
    auto drawSegment = [cr](const Point& p, const Point& q) {
        if (p.z == 0.0) return;
        cairo_set_line_width(cr, p.z);
        cairo_move_to(cr, p.x, p.y);
        cairo_line_to(cr, q.x, q.y);
        cairo_stroke(cr);
    };
```

**Comment by bhennion** [comment]
> Instead of asserting that the pressure is nonzero, maybe a simpler short-term solution is to skip drawing the segment if the first point is zero? Assuming no other invariants are broken, it should have the same effect.

Event simply removing the assert would work. Cairo is perfectly happy with null or even negative stroke widths. That's how things were before https://github.com/xournalpp/xournalpp/commit/553588a4cf858d76b357322e40f50f750ca69794.  But this just hides the issue rather than fixing it.

**Comment by Technius** [comment]
> Fixing opened corrupted file: we could say that if a point's pressure value is 0, we reset it to 0.5*(previousPoint.z + nextPoint.z) (if either previousPoint or nextPoint does not exist, just take half the value of the other one.

Here is how a stroke with some 0-pressure points is rendered in 1.1.2 (the example file edited so that some points in the middle have zero pressure):

![image](https://user-images.githubusercontent.com/1066652/197324279-73a3c22e-5986-4c2e-9d92-4ee32515ae9b.png)

The proposed interpolation fix will cause the "missing" parts of the stroke to suddenly appear. Some users may be surprised if their existing documents start looking different after they update Xournal++. I am not convinced that this is something we want to fix at all.

> If neither exist, delete the stroke).

This would cause the stroke in the image above to be deleted completely.

**Comment by bhennion** [comment]
Is this situation really possible!? The only instances of 0-width cases I saw were at the very end of a stroke, when the driver emits an event containing both BTN_TOUCH = 0 and ABS_PRESSURE = 0.
Of course, there must be drivers behaving in ways I can't predict :)

In any case, if your example actually happened IRL, then yes, I'd say this needs fixing (and splitting into two strokes). This could trigger bug reports like "the delete stroke eraser mode deletes strokes I do not touch".
Minor inconveniences, I agree, but still a faulty behaviour.

So, a better fix: upon file loading, when encountering a point with width == 0, end the stroke right here and htere, and parse the remaining points to see if there's something worth salvaging. If so, create another stroke (or more).

The question of negative width values is also due to an old bug (interpolating some values between the last point.z and Point::NO_PRESSURE = -1). This bug was resolved, so it's only about loading files properly. Probably the simplest solution is to take an absolute value upon loading. It would not change the stroke's appearence, I think.
